---
title: "BEAST preparation"
author: "Laura Kragh Frederiksen"
date: "2023-01-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:


## Loading tree, dropping tips and finding most recent common ancestor of fossil internodes.

Find nodes I want to use for calibration and run chronos.

```{r Read tree and find most recent common ancestor for the fossils.}
setwd("~/Documents/Uni - Biologi/Speciale/Cluster/Data_output/Final_tree/Astral")

library(phytools)
library(ape)
library(ggtree)

startingtree = read.tree(file='Beilschmiedia-Outgroup_trees_BP10_SpeciesTree_annotQ_rooted2.tre')

# Remove "" from species name.
startingtree$tip.label <- gsub("'","",as.character(startingtree$tip.label))
startingtree$tip.label <- gsub("pube-","puben",as.character(startingtree$tip.label))
startingtree[["tip.label"]]

tree <- read.tree(file='Beilschmiedia-Outgroup_trees_BP10_SpeciesTree_rooted2.tre')

ggtree(tree) + geom_text2(aes(subset=!isTip, label=node), hjust=-.3) + geom_tiplab()

# Dropping all tips of the nine species where I have multiple samples: Beil-manni-705AL1, Beil-manni706AL1, Pota-obtus-771AL1, Beil-tarai-722AL1, Beil-tawa-724AL1, Synd-marli-776AL1, Beil-latif-699AL1, Beil-roxbu-716AL1, Cryp-alba-800AL1, Cryp-litor-734AL1 and Cryp-polyn-782AL1.
startingtree <- drop.tip(phy = startingtree, tip = c(43,44,55,56,61,70,73,82,94,135,136))
startingtree[["tip.label"]]

# Write the ASTRAL tree to a file:
write.tree(startingtree, file="startingtree_noChronos")


# Fossil - Araria florifera, 113 Ma, stem Laurales. 
araria <- mrca(startingtree)['Myri-fragr-PAFTOL', 'Magn-grand-PAFTOL']

# Fossil - Virginianthus calycanthoides, 107.7 Ma, Crown Laurales.
virginianthus <- mrca(startingtree)['Idio-austr-PAFTOL', 'Sipa-guian-PAFTOL']

# Fossil - Potomacanthus lobatus, 106.8 Ma, Stem Lauraceae.
potomacanthus <- mrca(startingtree)['Cryp-alba-PAFTOL', 'Cass-filif-PAFTOL']

# Fossil - Jerseyanthus calycanthoides, 85.8 MA, Crown Calycanthoideae.
jerseyanthus <- mrca(startingtree)['Caly-flori-PAFTOL', 'Chim-salic-PAFTOL']

# Fossil - Neusenia tetrasporangiata, 83.6 Ma, Stem Neocinnamomum.
neusenia <- mrca(startingtree)['Neoc-cauda-PAFTOL', 'Cary-tonki-PAFTOL']

# Fossil - Monimiophyllum callidentatum, 52.22 Ma, Crown Monimiaceae.
monimiophyllum <- mrca(startingtree)['Peum-boldu-PAFTOL', 'Moll-gilgi-PAFTOL']

# Fossil - Illigera eocenica - 41.2 Ma, Crown Hernandiaceae.
illigera <- mrca(startingtree)['Hern-nymph-PAFTOL', 'Spar-botoc-PAFTOL']

# Fossil - Cryptocaryoxylon gippslandicum, 39 Ma, Crown clade C. aristata, C. oubatchensis, C. pluricostata and C. rhodosperma.
cryptocaryoxylon <- mrca(startingtree)['Cryp-alba-PAFTOL', 'Cryp-rhodo-784AL1']

# Fossil - Alseodaphne changchangensis - 37-48 Ma, Crown Persea group.
alseodaphne <- mrca(startingtree)['Phoe-lance-PAFTOL', 'Mach-salic-PAFTOL']

# Fossil - Machilus maomingensis - 33.7-33.9 Ma, Stem Machilus.
machilus <- mrca(startingtree)['Mach-salic-PAFTOL', 'Alse-petio-PAFTOL']

```

## Running chronos to get a calibrated starting tree and writing the starting tree to a file, ready for BEAUTi.

```{r Create startingtree to use in BEAST using the function chronos and the above nodes for calibration.}
setwd("~/Documents/Uni - Biologi/Speciale/Cluster/Data_output/Final_tree/Astral")

library(ape)

### I used this guide - https://justinbagley.rbind.io/2015/08/30/update-new-functions-for-generating-starting-trees-for-beast-or-starbeast-in-r/

# Placing calibration points using the command 'makeChronosCalib' in interactive mode embedded in the chronos function.
startingtree_Calib_cal_absolute <- makeChronosCalib(startingtree, node = c(156, 135, 222, 145, 139, 160, 146, 151, 149, 137), age.min = c(42.5, 113, 39, 41.2, 85.8, 33.8, 52.22, 83.6, 106.8, 107.7), interactive = FALSE, soft.bounds = FALSE)

# Making my 10-calibration time tree:
startingtree_Calib_chronogram <- chronos(startingtree, lambda = 0, model = 'relaxed', calibration = startingtree_Calib_cal_absolute)

# Plotting the new time tree:
plot.phylo(startingtree_Calib_chronogram, type = 'phylogram', use.edge.length=TRUE, font=3, cex=0.2, no.margin=TRUE)

# Write the time-calibrated starting tree to a file:
write.tree(startingtree_Calib_chronogram, file="startingtree_10fossils_relaxedclock")

# Write the time-calibrated starting tree to a file in Nexus format
write.nexus(startingtree_Calib_chronogram, file="startingtree_10fossils_relaxedclock_nexus")

```

## Manually preparing the 8 selected 'best' alignments for BEAUTi
In SublimeTeext I will manually remove the species that have multiple samples that I do not need, because it will become a problem when conducting the biogeographic analysis with BioGeoBEARS.

I am manually removing the following sequences Beil-manni-705AL1, Beil-manni706AL1, Pota-obtus-771AL1, Beil-tarai-722AL1, Beil-tawa-724AL1, Synd-marli-776AL1, Beil-latif-699AL1, Beil-roxbu-716AL1, Cryp-alba-800AL1, Cryp-litor-734AL1 and Cryp-polyn-782AL1.


## Insert taxa and gaps for species missing in each alignment so all alignments include the same equal amount of species
# This script is run through the cluster.

```{python Code to insert empty sequences}
# -*- coding: utf-8 -*-
###########################################################################
# Project: Beilschmiedia
# Script: add_empty_seqs_for_sp.py
# --- Action: Puts empty sequence for TAGs (individuals)
# ----------- that have no sequence (after alignment)
# ----------- This is necessary for linking tree files in the BEAUTi file before dating in BEAST.
# --- Input: alignments
# --- Output: alignments where empty sequences are added for TAGs which do not have alignment
# Author: Maya Schroedl
# Adapted by: Laura Kragh Frederiksen 
# The script is after adaptation compatible with biopyhton > v.1.78 where the Bio.Alphabet module is no longer included.

###########################################################################

#### Packages ####
import os
from Bio import SeqIO
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
import glob


##### DIRECTORIES #############
# Working directory
WD = os.getcwd() #project working directory

#Create directory "2_phylo_dating/1_alignment"
if not os.path.exists(WD+ "/2_phylo_dating/1_alignment"):
	    os.makedirs(WD+ "/2_phylo_dating/1_alignment")
	    
	    
#directory where alignments are
phylo_dir = WD+"/1_phylo_reconstruction/2_alignment/"
os.chdir(phylo_dir)

#get filenames of manually modified alignments (_clean)
#those are the final alignments
matching_filenames = glob.glob("*_clean.fasta") 

#for each alignment: add empty sequences for TAGs where no
for filename in matching_filenames:
    
    #open namelist (list of TAG names)
    with open(WD+"/1_phylo_reconstruction/namelist.txt","r") as totalsp:
        totalsp=totalsp.read().split('\n') #put into table
    
    #read the alignment which corresponds to the gene
    records = list(SeqIO.parse(WD+"/1_phylo_reconstruction/2_alignment/"+
                    filename, "fasta"))
    
    #TAG names of alignments
    recordids=[]
    
    #put names of TAGs which are in alignment to recordids
    for TAG in records:
        recordids.append(TAG.id)
        
    # get difference of the two lists: list of all TAGS and 
                                    #list of TAGs for which there is an alignment
    diff = list(set(totalsp) - set(recordids)) # = which TAGs need an empty sequence added
    
    # create an empty sequence with only "-" which has the length of the alignment
    sequence = Seq("-"*len(records[2].seq)) 
    
     # add sequence of "-" for all TAGs which are not represented in this gene alignment
    for TAG in diff:
        newrec = SeqRecord(sequence,id=TAG, description="") #new empty sequence with TAG id
        records.append(newrec)  
    
    # write new alignment (with added empty sequences) to the alignment folder in the dating folder
    SeqIO.write(records, WD+ "/2_phylo_dating/1_alignment/"+filename, "fasta")    
        

```


## Code to manipulate the wcvp dataframe
# Only include genera included in the ingroup.
The genera are Aspidostemon, Beilschmiedia, Cryptocarya, Endiandra, Eusideroxylon, Potameia, Potoxylon, Sinopora and Syndiclis.
```{r Code to manipulate the wcvp dataframe}
setwd("~/Documents/Uni - Biologi/Speciale/wcvp_distributionData")

# Read txt files:
wcvp_distribution <- read_delim("wcvp_distribution.txt", delim = "|", escape_double = FALSE, trim_ws = TRUE)

wcvp_names <- read_delim("wcvp_names.txt", delim = "|", escape_double = FALSE, trim_ws = TRUE)

# Join files by specific column - plant_name_id.
wcvp_combined <- merge(wcvp_names, wcvp_distribution, by=c("plant_name_id"), all = TRUE)

# Only keep rows from the family Lauraceae
wcvp_combined_oLauraceae <- filter(wcvp_combined, family == "Lauraceae")

# Only keep rows from the genera included in the ingroup in my tree.
wcvp_combined_oIngroup <- filter(wcvp_combined_oLauraceae, genus == "Aspidostemon" | genus ==  "Beilschmiedia" | genus =="Cryptocarya" | genus == "Endiandra" | genus == "Eusideroxylon" | genus == "Potameia" | genus == "Potoxylon" | genus == "Sinopora" | genus == "Syndiclis")

write_xlsx(x = wcvp_combined_oIngroup, "../wcvp_combined_oIngroup.xlsx")

```

## Visualise country data for ingroup genera on a world map.

```{r Visualise R wcvp data on a map}

# Read excel
wcvp_combined_oIngroup <- read_excel("wcvp_combined_oIngroup.xlsx")

ingroup.map <- joinCountryData2Map(wcvp_combined_oIngroup, joinCode = "NAME", nameJoinColumn = "country")

par(mar=c(0,0,1,0))
mapCountryData(ingroup.map, nameColumnToPlot="genus")

```




Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
