# Supplementary information for the article manuscript ["Long-distance dispersal shaped the pantropical distribution of the geographically structured Cryptocarya group"]() from the MSc thesis of Laura Kragh Frederiksen.

## DATA

**``Data_S1.zip``**

**Data S1: Alignments**
342 fasta files generated with MAFFT and trimmed with Trimal, Optrimal, CIalign and Taper. Each file represents one gene retrieved from the ['mega353' target file](https://github.com/chrisjackson-pellicle/NewTargets) (McLay et al., 2021). The faster headers correspond to each individual in short notation. See attached [text file]() for a conversion of the short notation to the species name.

**``Data_S2.zip``**

**Data S2: Gene trees**
342 ML (maximum likelihood) gene trees generated with IQtree in Newick format. Each gene tree corresponds to one gene. The faster headers correspond to each individual in short notation. See attached [text file]() for a conversion of the short notation to the species name.

## FIGURES

**``Figure_S1``**

**Figure S1: ** 

**``Figure_S2``** 

**Figure S2** 

**'Figure_S3'**

**Figure S3**

**'Figure_S4'**

**Figure S4**

Figure S5

Figure S6

## TABLES

**``Table_S1.xlsx``**

**Table S1: Gene tree concordance to 'PAFTOL' outgroup topology**
Nuclear [Angiosperms353](https://www.ncbi.nlm.nih.gov/pubmed/30535394) (Johnson et al., 2019) genes used to generate the [PAFTOL](https://www.ncbi.nlm.nih.gov/pubmed/33983440) (Baker et al., 2022) outgroup topology. We used [SortaDate](https://www.ncbi.nlm.nih.gov/pubmed/29772020) (Smith et al., 2018) to identify the supercontig genes that were most concordant to the PAFTOL outgroup topology based on exons (Baker et al., 2022). We then tested how many exons we were able to substitute with supercontigs without losing the desired topology used in [Ramirez-Barahona (2020)](https://www.nature.com/articles/s41559-020-1241-3). This was 10%. The genes in the supercontig table are the 10% most concordant genes, whereas the genes in the exon table are the remaining 90 %.

**``Table_S2.xlsx``**

**Table S2: HybPiper recovery stats**
[HybPiper](https://bioone.org/journals/applications-in-plant-sciences/volume-4/issue-7/apps.1600016/HybPiper--Extracting-Coding-Sequence-and-Introns-for-Phylogenetics-from/10.3732/apps.1600016.full) target enrichment and gene recovery efficiency summary for the 115 ingroup samples (*Syndiclis chinensis* included after ['coverage']() step) and 26 outgroup samples (*Peumus boldus* and *Chimonanthus salicifolia* included after ['coverage']() step). The columns in the table correspond to: Sample name , Number of reads,	Number of reads on target, Percentage of reads on target,	Number of genes with reads,	Number of genes with contigs,	Number of genes with sequences,	Number of genes with sequences > 25% of the target length, Number of genes with sequences > 50% of the target length, Number of genes with sequences > 75% of the target length, Number of genes with sequences < 150% of the target length, Number of genes with paralog warnings. See [HybPiper wiki](https://github.com/mossmatters/HybPiper/wiki#hybpiper-stats) for an explanation of what the different outputs mean. See attached [text file]() for a conversion of the short notation to the species name.

# Table S3

## TEXT

**``Text_S1.pdf``**

**Text S1: Fossil selection**
Justification for the eight fossils included as calibration points in our molecular dating with [BEAST]() () based on [Ramirez-Barahona et al. (2020)](https://www.nature.com/articles/s41559-020-1241-3) and [Li et al. (2020)](https://www.ncbi.nlm.nih.gov/pubmed/32619568).
