# Dating the phylogeny with BEAST

[BEAST](https://beast.community) [(Drummond & Rambaut, 2007)](https://www.ncbi.nlm.nih.gov/pubmed/17996036
https://bmcecolevol.biomedcentral.com/counter/pdf/10.1186/1471-2148-7-214.pdf) based on Bayesian inference was used to date the phylogenetic tree. To do this we used the three best nuclear genes, nine fossils (primary calibrations) and one secondary calibration on the root (see Table ... in [MSc thesis]()). Moreover, we fixed the tree topology to that of the ASTRAL tree based on 34. nuclear genes. 

We selected the fossils based on [Ramirez-Barahona et al., 2022](https://www.nature.com/articles/s41559-020-1241-3) and the fossils included in [Li et al. (2020)](https://www.ncbi.nlm.nih.gov/pubmed/32619568), except the fossil [_Cryptocaryoxylon gippslandicum_](https://doi.org/10.1080/03115518608619157).

To see the 

## Selecting genes

To select genes for dating I used [SortaData](https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0197433&type=printable) to get a list of the 'best- genes, thus sorted after the genes with least root to tip variance and best bipartition support compared to the species tree.  

## Alignments

## Preparing the xml file in BEAUti

## Running BEAST


